pkg/descheduler/strategies/node_selector_test.go

```diff
@@ -0,0 +1,153 @@
+	ctx := context.Background()
+	selectorStrategy := api.DeschedulerStrategy{
+		Enabled: true,
+	}
+
+	selectorStrategyWithDegradation := api.DeschedulerStrategy{
+		Enabled: true,
+		Params: &api.StrategyParameters{
+			NodeSelection: &api.NodeSelection{
+				DegradationAllowed: true,
+			},
+		},
+	}
+
+	nodeLabelKey := "kubernetes.io/desiredNode"
+	incompatibleNodeLabelKey := "kubernetes.io/undesiredNode"
+	nodeLabelValue := "yes"
+	nodeWithLabels := test.BuildTestNode("nodeWithLabels", 2000, 3000, 10, nil)
+	nodeWithLabels.Labels[nodeLabelKey] = nodeLabelValue
+
+	nodeWithoutLabels := test.BuildTestNode("nodeWithoutLabels", 2000, 3000, 10, nil)
+
+	unschedulableNodeWithLabels := test.BuildTestNode("unschedulableNodeWithLabels", 2000, 3000, 10, nil)
+	nodeWithLabels.Labels[nodeLabelKey] = nodeLabelValue
+	unschedulableNodeWithLabels.Spec.Unschedulable = true
+
+	addPodsToNode := func(node *v1.Node) []v1.Pod {
+		podWithNodeSelector := test.BuildTestPod("podWithNodeSelector", 100, 0, node.Name, nil)
+		podWithNodeSelector.Spec.NodeSelector = map[string]string{
+			nodeLabelKey: nodeLabelValue,
+		}
+
+		podWithIncompatibleNodeSelector := test.BuildTestPod("podWithIncompatibleNodeSelector", 100, 0, node.Name, nil)
+		podWithIncompatibleNodeSelector.Spec.NodeSelector = map[string]string{
+			incompatibleNodeLabelKey: nodeLabelKey,
+		}
+
+		unlabelledPod := test.BuildTestPod("unlabelledPod", 100, 0, node.Name, nil)
+
+		podWithNodeSelector.ObjectMeta.OwnerReferences = test.GetNormalPodOwnerRefList()
+		podWithIncompatibleNodeSelector.ObjectMeta.OwnerReferences = test.GetNormalPodOwnerRefList()
+		unlabelledPod.ObjectMeta.OwnerReferences = test.GetNormalPodOwnerRefList()
+
+		return []v1.Pod{
+			*podWithNodeSelector,
+			*podWithIncompatibleNodeSelector,
+			*unlabelledPod,
+		}
+	}
+
+	tests := []struct {
+		description             string
+		nodes                   []*v1.Node
+		pods                    []v1.Pod
+		strategy                api.DeschedulerStrategy
+		expectedEvictedPodCount int
+		maxPodsToEvict          int
+	}{
+		{
+			description:             "Pod is correctly scheduled on node, no eviction expected",
+			strategy:                selectorStrategy,
+			expectedEvictedPodCount: 0,
+			pods:                    addPodsToNode(nodeWithLabels),
+			nodes:                   []*v1.Node{nodeWithLabels},
+			maxPodsToEvict:          0,
+		},
+		{
+			description:             "Pod is scheduled on node without matching labels, another schedulable node available, should be evicted",
+			expectedEvictedPodCount: 1,
+			strategy:                selectorStrategy,
+			pods:                    addPodsToNode(nodeWithoutLabels),
+			nodes:                   []*v1.Node{nodeWithoutLabels, nodeWithLabels},
+			maxPodsToEvict:          0,
+		},
+		{
+			description:             "Pod is scheduled on node without matching labels, another schedulable node available, maxPodsToEvict set to 1, should not be evicted",
+			expectedEvictedPodCount: 1,
```

Comment:
`Pod is scheduled on node without matching labels, another schedulable node available, maxPodsToEvict set to 1, should not be evicted` contradicts `expectedEvictedPodCount` set to `1`.